app/controllers/spree/khalti_controller.rb

```diff
@@ -0,0 +1,116 @@
+module Spree
+  class KhaltiController < StoreController
+    skip_before_action :verify_authenticity_token
+
+    def payment
+      payment_method = PaymentMethod.find(params[:payment_method_id])
+
+      order = current_order || raise(ActiveRecord::RecordNotFound)
+
+      current_payment = create_payment(payment_method)
+      current_payment.pend!
+
+      begin
+        # API call to khalti
+        headers = {
+          Authorization: "Key #{payment_method.preferences[:test_secret_key]}"
+        }
+        uri = URI.parse('https://khalti.com/api/v2/payment/verify/')
+        https = Net::HTTP.new(uri.host, uri.port)
+        https.use_ssl = true
+        request = Net::HTTP::Post.new(uri.request_uri, headers)
+        request.set_form_data('token' => params[:payload][:token], 'amount' => params[:payload][:amount].to_f)
+        response = https.request(request)
+
+        if response_code.eql?(200)
+          responseJSON = JSON.parse(response.body)
+
+          # Update the payment source with the response then complete
+          current_payment.source.update!(khalti_response_attributes(responseJSON).merge({payment_method_id: payment_method.id, user_id: current_order.user.id}))
+          current_payment.complete!
+
+          unless current_order.next
+            flash[:error] = @order.errors.full_messages.join("\n")
+            redirect_to checkout_state_path(current_order.state) and return
+          end
+
+          redirect_to completion_route(order)
+        else
+          current_payment.invalid!
+          flash[:error] = Spree.t('flash.generic_error', scope: 'khalti', reasons: 'Server verification with khalti failed')
+          redirect_to checkout_state_path(:payment)
+        end
+      rescue SocketError
+        current_payment.invalid!
+        flash[:error] = Spree.t('flash.connection_failed', scope: 'khalti')
+        redirect_to checkout_state_path(:payment)
+      end
+
+    end
+
+    # To be used when configurations need to be fetched from the UI
+    def khalti_payment_config
+      @payment_method = PaymentMethod.find(params[:payment_method_id])
+      config = {
+        publicKey: @payment_method.preferences[:test_public_key],
+        productIdentity: current_order.number,
+        productName: current_order.number,
+        productUrl: "#{current_store.url}/orders/#{current_order.number}",
+        paymnetPreference: ["MOBILE_BANKING", "KHALTI", "EBANKING","CONNECT_IPS","SCT"],
```

Comment:
`paymnetPreference` seems like type, is this what Khalti demand?

Reply:
This is what khalti demands, through currently we are not using this configuration from here and there is a typo. Will fix the typo.